Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs

```diff
@@ -0,0 +1,238 @@
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallMissing:
+                            ExecNpmInstallMissing(node);
+                            return VSConstants.S_OK;
+
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallNew:
+                            ExecNpmInstallNew(node);
+                            return VSConstants.S_OK;
+
+                        case PkgCmdId.cmdidWorkSpaceNpmUpdate:
+                            ExecNpmUpdate(node);
+                            return VSConstants.S_OK;
+                    }
+
+                    if (nCmdID >= PkgCmdId.cmdidWorkSpaceNpmDynamicScript && nCmdID < PkgCmdId.cmdidWorkSpaceNpmDynamicScriptMax)
+                    {
+                        ExecDynammic(node, nCmdID);
+                        return VSConstants.S_OK;
+                    }
+                }
+
+                return (int)Constants.OLECMDERR_E_NOTSUPPORTED;
+            }
+
+            // Note: all the Exec commands are async, this allows us to call them in a fire and forget
+            // pattern, without blocking the UI or losing any logging
+
+            private async void ExecNpmInstallMissing(WorkspaceVisualNodeBase node)
+            {
+                using (var npmController = this.CreateController(node.Workspace))
+                using (var commander = npmController.CreateNpmCommander())
+                {
+                    await commander.Install();
+                }
+            }
+
+            private void ExecNpmInstallNew(WorkspaceVisualNodeBase node)
+            {
+                using (var npmController = this.CreateController(node.Workspace))
+                using (var npmWorker = new NpmWorker(npmController))
+                using (var manager = new NpmPackageInstallWindow(npmController, npmWorker))
+                {
+                    manager.ShowModal();
+                }
+            }
+
+            private async void ExecNpmUpdate(WorkspaceVisualNodeBase node)
+            {
+                using (var npmController = this.CreateController(node.Workspace))
+                using (var commander = npmController.CreateNpmCommander())
+                {
+                    await commander.UpdatePackagesAsync();
+                }
+            }
+
+            private async void ExecDynammic(WorkspaceVisualNodeBase node, uint nCmdID)
+            {
+                // Unfortunately the NpmController (and NpmCommander), used for the install, update commands
+                // doesn't support running arbitrary scripts. And changing that is outside
+                // the scope of these changes.
+
+                var index = nCmdID - PkgCmdId.cmdidWorkSpaceNpmDynamicScript;
+                var packageJson = PackageJsonFactory.Create(((IFileNode)node).FullPath);
+
+                Debug.Assert(packageJson != null, "Failed to create package.json");
+
+                var scripts = packageJson.Scripts;
+                if (index < scripts.Length)
```

Comment:
>if (index < scripts.Length) [](start = 16, length = 27)

Can this fail?

Reply:
Generally no, unless an external process changes the `package.json` while the menu is open.... e..g git pull

Reply:
Um, shouldn't we ignore *any* index if that happens?  What if someone deletes script 0?  The user probably doesn't want the script after the one they clicked to run.

Reply:
yeah, but considering how unlikely this scenario is...